sjsonnet/src/sjsonnet/Importer.scala

```diff
@@ -1,37 +1,172 @@
 
-import fastparse.Parsed
 
 /** Resolve and read imported files */
 abstract class Importer {
   def resolve(docBase: Path, importName: String): Option[Path]
-  def read(path: Path): Option[String]
+  def read(path: Path): Option[ResolvedFile]
 
-  def resolveAndRead(docBase: Path, importName: String): Option[(Path, String)] = for {
+  def resolveAndRead(docBase: Path, importName: String): Option[(Path, ResolvedFile)] = for {
     path <- resolve(docBase, importName)
     txt <- read(path)
   } yield (path, txt)
 
-  def resolveAndReadOrFail(value: String, pos: Position)(implicit ev: EvalErrorScope): (Path, String) =
+  def resolveAndReadOrFail(value: String, pos: Position)(implicit ev: EvalErrorScope): (Path, ResolvedFile) =
     resolveAndRead(pos.fileScope.currentFile.parent(), value)
       .getOrElse(Error.fail("Couldn't import file: " + pprint.Util.literalize(value), pos))
 }
 
 object Importer {
   val empty: Importer = new Importer {
     def resolve(docBase: Path, importName: String): Option[Path] = None
-    def read(path: Path): Option[String] = None
+    def read(path: Path): Option[ResolvedFile] = None
+  }
+}
+
+case class FileParserInput(file: File) extends ParserInput {
+
+  private[this] val bufferedFile = new BufferedRandomAccessFile(file.getAbsolutePath, 1024 * 8)
+
+  private lazy val fileLength = file.length.toInt
+
+  override def apply(index: Int): Char = {
+    bufferedFile.readChar(index)
+  }
+
+  override def dropBuffer(index: Int): Unit = {}
+
+  override def slice(from: Int, until: Int): String = {
+    bufferedFile.readString(from, until)
+  }
+
+  override def length: Int = fileLength
+
+  override def innerLength: Int = length
+
+  override def isReachable(index: Int): Boolean = index < length
+
+  override def checkTraceable(): Unit = {}
+
+  private[this] lazy val lineNumberLookup: Array[Int] = {
+    val lines = mutable.ArrayBuffer[Int]()
+    val bufferedStream = new BufferedInputStream(new FileInputStream(file))
+    var byteRead: Int = 0
+    var currentPosition = 0
+
+    while ({ byteRead = bufferedStream.read(); byteRead != -1 }) {
+      if (byteRead == '\n') {
+        lines += currentPosition + 1
+      }
+      currentPosition += 1
+    }
+
+    bufferedStream.close()
+
+    lines.toArray
+  }
+
+  def prettyIndex(index: Int): String = {
+    val line = lineNumberLookup.indexWhere(_ > index) match {
+      case -1 => lineNumberLookup.length - 1
+      case n => math.max(0, n - 1)
+    }
+    val col = index - lineNumberLookup(line)
+    s"${line + 1}:${col + 1}"
   }
 }
 
+class BufferedRandomAccessFile(fileName: String, bufferSize: Int) {
+
+  // The file is opened in read-only mode
+  private val file = new RandomAccessFile(fileName, "r")
+
+  private val buffer = new Array[Byte](bufferSize)
+
+  private var bufferStart: Long = -1
+
+  private var bufferEnd: Long = -1
+
+  private val fileLength: Long = file.length()
+
+  private def fillBuffer(position: Long): Unit = {
+    file.seek(position)
```

Comment:
Do we always need to do a `file.seek` here? I imagine in the common case that `fillBuffer` gets called, `position` would simply be equivalent to `bufferEnd`, in which case the `seek` is unnecessary. Leaving it out would halve the number of filesystem calls.

Reply:
Good point, added a check. If we're already at the right position in the file, we skip the seek.